tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/LicenseChecker.java

```diff
@@ -0,0 +1,329 @@
+		Set<IncludedDependency> declaredDependencies = new HashSet<>();
+		for (String line : noticeContents) {
+			Matcher noticeDependencyMatcher = NOTICE_DEPENDENCY_PATTERN.matcher(line);
+			if (noticeDependencyMatcher.find()) {
+				String groupId = noticeDependencyMatcher.group(1);
+				String artifactId = noticeDependencyMatcher.group(2);
+				String version = noticeDependencyMatcher.group(3);
+				if (groupId == null && artifactId == null && version == null) { // "bundles" case
+					groupId = noticeDependencyMatcher.group(5);
+					artifactId = noticeDependencyMatcher.group(6);
+					version = noticeDependencyMatcher.group(7);
+				}
+				IncludedDependency toAdd = IncludedDependency.create(groupId, artifactId, version);
+				if (!declaredDependencies.add(toAdd)) {
+					LOG.error("Dependency {} has been declared twice in module {}", toAdd, moduleName);
+					severeIssueCount++;
+				}
+			}
+		}
+		// print all dependencies missing from NOTICE file
+		Collection<IncludedDependency> expectedDependencies = modulesWithShadedDependencies.get(moduleName);
+		for (IncludedDependency expectedDependency : expectedDependencies) {
+			if (!declaredDependencies.contains(expectedDependency)) {
+				LOG.error("Could not find dependency {} in NOTICE file {}", expectedDependency, noticeFile);
+				severeIssueCount++;
+			}
+		}
+
+		boolean moduleDefinesExcessDependencies = MODULES_DEFINING_EXCESS_DEPENDENCIES.contains(moduleName);
+
+		// print all dependencies defined in NOTICE file, which were not expected
+		for (IncludedDependency declaredDependency : declaredDependencies) {
+			if (!expectedDependencies.contains(declaredDependency)) {
+				if (moduleDefinesExcessDependencies) {
+					LOG.debug(
+						"Dependency {} is mentioned in NOTICE file {}, but was not mentioned by the build output as a bundled dependency",
+						declaredDependency,
+						noticeFile);
+				} else {
+					LOG.warn(
+						"Dependency {} is mentioned in NOTICE file {}, but is not expected there",
+						declaredDependency,
+						noticeFile);
+				}
+
+			}
+		}
+
+		return severeIssueCount;
+	}
+
+	private static List<Path> findNoticeFiles(Path root) throws IOException {
+		return Files.walk(root)
+			.filter(file -> {
+				int nameCount = file.getNameCount();
+				return file.getName(nameCount - 3).toString().equals("resources")
+					&& file.getName(nameCount - 2).toString().equals("META-INF")
+					&& file.getName(nameCount - 1).toString().equals("NOTICE");
+			})
+			.collect(Collectors.toList());
+	}
+
+	private static Multimap<String, IncludedDependency> parseModulesFromBuildResult(File buildResult) throws IOException {
+		Multimap<String, IncludedDependency> result = ArrayListMultimap.create();
+
+		try (Stream<String> lines = Files.lines(buildResult.toPath())) {
+			//String line;
+			String currentShadeModule = null;
+			String currentDependencyCopyModule = null;
+			for (String line : (Iterable<String>) lines::iterator) {
```

Comment:
let's do it like this then (+ my comment below)

```suggestion
			String currentShadeModule = null;
			String currentDependencyCopyModule = null;
			lines.forEach(line -> {
```

Reply:
yeah, but I can not modify the `currentShadeModule` variables then. Of course I could put them into some container, but that would be ugly.

Reply:
I'm really not sure if it's worth going into this rabbit hole for a tiny syntactic improvement in an internal utility

Reply:
damn it. I just dislike the cast.

Reply:
but let's leave it like it is